agent/boundarylogproxy/codec/codec.go

```diff
@@ -0,0 +1,102 @@
+// Package codec implements the wire format for agent <-> boundary communication.
+//
+// Wire Format:
+//   - 4 bits: big-endian tag
+//   - 28 bits: big-endian length of the protobuf data
+//   - length bytes: encoded protobuf data
+//
+// Note that while there are 28 bits for the length, in practice the messages should
+// be much smaller. 28 bits is chosen because sizing down the header to 16 bits would
+// sacrifice future flexibility. For example, a 3 bit tag and 13 bit length doesn't
+// have much length headroom should the batching strategy change.
+package codec
+
+import (
+	"encoding/binary"
+	"io"
+
+	"golang.org/x/xerrors"
+)
+
+type Tag uint8
+
+const (
+	// TagV1 identifies the first revision of the protocol.
+	TagV1 Tag = 1
+)
+
+const (
+	// DataLength is the number of bits used for the length of encoded protobuf data.
+	DataLength = 28
+
+	// tagLength is the number of bits used for the tag.
+	tagLength = 4
+
+	// MaxMessageSize is the practical maximum size of the protobuf messages
+	// sent over the wire. While the wire format allows for messages much larger
+	// than this, practically they are not expected to be.
+	MaxMessageSize = 1 << 15
+)
+
+var ErrTooLarge = xerrors.New("message too large")
+
+// WriteFrame writes a framed message with the given tag and data. The tag must
+// fit in 4 bits (0-15), and data must not exceed 2^DataLength in length.
+func WriteFrame(w io.Writer, tag Tag, data []byte) error {
+	if len(data) > 1<<DataLength {
+		return xerrors.Errorf("data too large: %d bytes", len(data))
+	}
+
+	var header uint32
+	//nolint:gosec // The length check above ensures there's no overflow.
+	header |= uint32(len(data))
+	header |= uint32(tag) << DataLength
+
+	if err := binary.Write(w, binary.BigEndian, header); err != nil {
+		return err
+	}
+	_, err := w.Write(data)
+	return err
+}
+
+// ReadFrame reads a framed message, returning the decoded tag and data. If the
+// message size exceeds maxSize, ErrTooLarge is returned. The provided buf is
+// used if it has sufficient capacity; otherwise a new buffer is allocated. To
+// reuse the buffer across calls, pass in the returned data slice:
+//
+//	buf := make([]byte, initialSize)
+//	for {
+//	    _, buf, _ = ReadFrame(r, maxSize, buf)
+//	}
+func ReadFrame(r io.Reader, maxSize uint32, buf []byte) (Tag, []byte, error) {
```

Comment:
I assume you're allowing `maxSize` to be passed in here so that it can be increased in the future without API changes?

The awkward thing about this is that you can write 256MiB to the server but you're only allowed to read 32KiB. Both ends should agree on the allowable length.

What about enforcing the limit in both read/write transparently (i.e. without accepting a length arg), and bumping the version if we increase this number? That way we can already prevent payloads from being written in the first place before we read them, to defend the server.